### PR TITLE
Add support for Windows on Arm

### PR DESCRIPTION
**Pull request by pbo-linaro**
https://github.com/osquery/osquery/issues/7901

This PR implements support for Windows on Arm, and CI using cross-compilation.
Some limitations are present, details are in commit messages.

- [X] Read the `CONTRIBUTING.md` guide at the root of the repo.
- [X] Ensure the code is formatted building the `format_check` target.  
      If it is not, then move the committed files to the git staging area,
      build the `format` target to format them, and then re-commit.
      [More information is available on the wiki](https://osquery.readthedocs.io/en/latest/development/building/#formatting-the-code).
- [X] Ensure your PR contains a single logical change.
- [ ] Ensure your PR contains tests for the changes you're submitting.
- [X] Describe your changes with as much detail as you can.
- [X] Link any issues this PR is related to.
- [X] Remove the text above.

### CODE REVIEW

**Comment by pbo-linaro** [comment]
Opened this PR as there were no answers on the original issue.

File: cmake/globals.cmake [hunk 1]
```diff
@@ -36,6 +36,9 @@ elseif(CMAKE_SYSTEM_PROCESSOR STREQUAL "aarch64")
 elseif(CMAKE_SYSTEM_PROCESSOR STREQUAL "arm64")
   # Apple Silicon
   set(TARGET_PROCESSOR "aarch64")
+elseif(CMAKE_SYSTEM_PROCESSOR STREQUAL "ARM64")
+  # Windows on Arm
+  set(TARGET_PROCESSOR "aarch64")
```
Comment (directionless) [inline cmake/globals.cmake:41]: These only differ by case is there a better way to handle this?

**Review by directionless — COMMENTED** [review]
Pretty neat to see this! @Smjert or @alessandrogario how's this look to you?

One thing we probably need to figure out -- how does this integrate into the builds and test cycle? I imagine it's going to be a bit like our other arm64 platforms.

File: cmake/globals.cmake (hunk 1 above)
Comment (pbo-linaro) [reply cmake/globals.cmake:41]: In general, I prefer to use explicit code for something like this, but it's possible to use [TOLOWER](https://cmake.org/cmake/help/latest/command/string.html#tolower) function instead.

**Comment by pbo-linaro** [comment]
IMHO, the main issue is the lack of support for AWS sdk, and it's not gonna be fixed soon.

For building, it should be possible to cross compile.

For testing, a dedicated Windows on Arm machine (or Virtual machine) is needed. If you already have a linux arm64 or Apple M1 machine somewhere, it should be possible to deploy a Win11 Arm virtual machine running on this.

**Comment by pbo-linaro** [comment]
Hi, polite ping :)

**Review by directionless — COMMENTED** [review]
We talked about this a bit in office hours today (2023-02-14). Bringing in new platforms can be surprisingly hard -- it incurs a lot ongoing maintenance cost. One thing that can help, is creating a CI job for this.

We think this should be able to be cross compiled from the windows 64bit workers. Is that something you can plumb into the setup? I think it would be a bit like the windows 32bit CI jobs

**Comment by pbo-linaro** [comment]
Yes, sure, we can understand this. I tried cross compiling from an x64 machine, and I ran into some issues, related to cmake. I think it's possible to solve it, but it would take some time. Same to add support in your CI.

Meanwhile, do you think we could first merge that PR, and then, come back to add cross-compilation support in another PR? We can easily run a nightly CI on our side (Linaro), and fix any regression that might happen for the windows-arm64, so this won't be broken.

At least, with this, people could build from source easily.

**Comment by estumpfo** [comment]
I'm a Product Manager in the Cisco Security team, and I just wanted to highlight that osquery support for Windows on Arm is something that many Cisco customers are asking for. 

For context, Cisco offers a solution that leverages osquery ([Cisco Orbital](https://orbital.amp.cisco.com/help/what-is-orbital/)) with wide adoption among Security Practitioners. As Windows on Arm grows in popularity, I've seen increased interest in Windows on Arm and think that many organizations would benefit from covering this gap.

**Comment by pbo-linaro** [comment]
@directionless After discussion, I will work on adding a cross compilation job in existing CI for OSQuery, so you'll feel more confident on integrating this :+1:

**Comment by pbo-linaro** [comment]
I just updated this PR, by adding CI support for Windows on Arm. You can find a successful pipeline on my [fork](https://github.com/pbo-linaro/osquery/actions/runs/4372912196).

There is one extra commit (not included in this PR), as I had to disable other platforms to save my GH Actions minutes, and avoid running self-hosted stuff I don't have access to.

---

By rebasing I ran into an issue with OpenSSL 1.1.1t update. For windows-arm64, we have to stay (until `u` version) to `s` version. You can find details in the concerned commit.

In more, only zip package is created for now. I can definitely look at integrating arm64 in osquery-packaging repo (for msi, nuget support), but if possible, that would be nice to have already this upstream :+1: 

Thanks

**Comment by pbo-linaro** [comment]
@directionless Polite ping after a week :)

Is there something more to help integrating this new platform?
Cisco needs this, and we are willing to help with support for this.

**Comment by pbo-linaro** [comment]
Just for information, expect build times similar to windows-x64 platform, without the testing step. An example pipeline (ran on Github actions runners) is available in my previous message.

**Comment by marcosd4h** [comment]
@pbo-linaro I'm looking into this and would like to test the CI-generated binaries. Would it be possible for you to upload the binaries somewhere so I can use that for testing? Thanks!

**Comment by Smjert** [comment]
Thanks @marcosd4h!
i don't want to cause PR fatigue, because my intervention is quite late, but a heads up that the libraries that Windows ARM64 uses are more than what we have a config here for.
We always create a hardcoded configuration and some notes on the steps taken for each of the libraries that a platform uses, not only the ones that immediately cause a compiler error, because some issues are not seen at compile time.
But also simply because if we have to update one of the libraries that has not been dealt with, we have to start from scratch.

**Comment by pbo-linaro** [comment]
> @pbo-linaro I'm looking into this and would like to test the CI-generated binaries. Would it be possible for you to upload the binaries somewhere so I can use that for testing? Thanks!

A CI run can be found [here](https://github.com/pbo-linaro/osquery/actions/runs/4372912196), including built binaries (see bottom of page).

**Comment by pbo-linaro** [comment]
@Smjert I'm willing to complete the patch as needed. Would that be possible for you to take some time to detail exact list of dependencies missing cmake config files?

Concerning PR fatigue, having feedback in a "not reactive" delay asking for more work is never a problem. We all are busy, it's fine. However, asking something and never answering back is a bit more tough IMHO.

**Comment by Smjert** [comment]
> @Smjert I'm willing to complete the patch as needed. Would that be possible for you to take some time to detail exact list of dependencies missing cmake config files?
> 
I was more referring to the config file that normally gets generated by the original library build system.
As far as I can see, if I didn't not miss anything, `librdkafka` is the only one: https://github.com/osquery/osquery/tree/master/libraries/cmake/source/librdkafka/config

In general, the list of libraries per platform is here (and mind that some are added just below, behind conditions): https://github.com/osquery/osquery/blob/6ac4acb45bd801f196ca923a5350fa7d15f3a5b1/CMakeLists.txt#L88

Finally, each library should have a README.md where we have notes for how to configure and build with the original build system, given that some options/features that are enabled or disabled are not part of the default config of the original build system. This helps reproducing the outputs and also helps when the library has to be updated.

**Comment by pbo-linaro** [comment]
Thanks for the details @Smjert.

I'll update the PR next week with the missing configuration file.
If there's anything else to modify to be able to merge this, you're welcome :)

**Comment by Smjert** [comment]
> Thanks for the details @Smjert.
> 
> I'll update the PR next week with the missing configuration file. If there's anything else to modify to be able to merge this, you're welcome :)

As I was mentioning, if possible also update the READMEs with any different flag/option you had to choose to configure with the original build system.
For instance, with `libarchive` we have https://github.com/osquery/osquery/tree/master/libraries/cmake/source/libarchive/README.md, at them bottom, if you've used the command as is, maybe at least add next to the `Windows` title, `(x86_64, ARM)`, so we know it has been used as is for both.
Things like this that will simplify work for who has to update and will increase the chance that the build is kept working, keeping in mind that it will be hard without being able to test.

**Comment by pbo-linaro** [comment]
No problem, I'll add this information too!

**Comment by pbo-linaro** [comment]
Updated PR as requested:
- Add missing config files for librdkafka
- Ensure all README for dependencies describe how to get config files for windows-arm64

I regenerated all config files to ensure they match exactly what is expected and are up to date :+1: 

@directionless @Smjert 
Would that be enough to accept to merge this new platform in OSQuery?

Thanks,
Pierrick

**Comment by directionless** [comment]
Sounds like this is on @Smjert, and he's a bit underwater. He'll try to peek this week

**Comment by pbo-linaro** [comment]
Thanks @directionless! We are committed to upstream this and help to maintain it after, for the windows-arm64 part. Would be nice to see it officially supported for next release of osquery.

File: libraries/cmake/formula/openssl/CMakeLists.txt [hunk 2]
```diff
@@ -3,6 +3,14 @@ project(thirdparty_openssl)
 set(OPENSSL_VERSION "1.1.1t")
 set(OPENSSL_ARCHIVE_SHA256 "8dee9b24bdb1dcbf0c3d1e9b02fb8f6bf22165e807f45adeb7c9677536859d3b")
 
+if(DEFINED PLATFORM_WINDOWS AND "${CMAKE_SYSTEM_PROCESSOR}" STREQUAL "ARM64")
+    # 1.1.1t is broken for windows-arm64 (between 1.1.1s and 1.1.1t, from 43d8f88)
+    # Following commit fixes it, but will be available only in 1.1.1u:
+    # https://github.com/openssl/openssl/commit/eec34033552c810132736c772d37547f097a
+    set(OPENSSL_VERSION "1.1.1s")
+    set(OPENSSL_ARCHIVE_SHA256 "c5ac01e760ee6ff0dab61d6b2bbd30146724d063eb322180c6f18a6f74e4b6aa")
+endif()
```
Comment (Smjert) [inline libraries/cmake/formula/openssl/CMakeLists.txt:12]: We can address this in a later PR, but I think we should download 1.1.1t and then apply a patch with the fix for the build that 1.1.1u has.
We can apply patches using ExternalProject in CMake: https://cmake.org/cmake/help/latest/module/ExternalProject.html (PATCH_COMMAND)

File: .github/workflows/hosted_runners.yml [hunk 3]
```diff
@@ -1196,7 +1217,12 @@ jobs:
       working-directory: ${{ steps.build_paths.outputs.BINARY }}
       shell: cmd
       run: |
-        call "C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Auxiliary\Build\vcvars${{ matrix.bitness }}.bat"
+        if ${{ matrix.bitness }}==arm64 (
+           echo Arm64 binaries can't be run on x64
+           exit 0
+        )
```
Comment (Smjert) [inline .github/workflows/hosted_runners.yml:1223]: nit: (Can clean up later if you'd like) lets just use the Github Action `if` as for the MSI case.

**Review by Smjert — APPROVED** [review]
Thanks for all the work here!
I'll open an issue to track the openssl update for Windows on ARM, but otherwise looks good.

File: libraries/cmake/formula/openssl/CMakeLists.txt (hunk 2 above)
Comment (pbo-linaro) [reply libraries/cmake/formula/openssl/CMakeLists.txt:12]: I'll update this PR with this, no problem.

File: .github/workflows/hosted_runners.yml (hunk 3 above)
Comment (pbo-linaro) [reply .github/workflows/hosted_runners.yml:1223]: Will clean up this in this PR

**Comment by pbo-linaro** [comment]
Added patch for OpenSSL (instead of using specific version for windows-arm64), and changed CI test job to be disabled for this platform using Github Actions "if".

Is is in a good state to be merged now?

**Review by Smjert — APPROVED** [review]

**Comment by pbo-linaro** [comment]
Thanks for your support on merging this.
Looking forward for official binaries with next release of osquery!

**Comment by directionless** [comment]
@pbo-linaro Thank you! It's a great contribution, and I'm really glad to see osquery on additional platforms

**Comment by lszcs90** [comment]
Hello,

I tried to build for ARM64 using tag 5.9.1 with: 
`cmake -G "Visual Studio 17 2022" -A arm64 .. -DOSQUERY_BUILD_AWS=OFF`  
(created a folder for build of course)
and after that running 
`cmake -G "Visual Studio 17 2022" -A arm64 .. -DOSQUERY_BUILD_AWS=OFF`
or just opening sln with visual studio and running a build on "ALL_BUILD". 

I got couple of errors: 
1. project thirdparty_boost_context: 
3>D:\Microsoft Visual Studio\2022\Professional\MSBuild\Microsoft\VC\v170\BuildCustomizations\masm.targets(70,5): error MSB3721: The command "echo MASM not supported on this platform
3>D:\Microsoft Visual Studio\2022\Professional\MSBuild\Microsoft\VC\v170\BuildCustomizations\masm.targets(70,5): error MSB3721: exit 1" exited with code 1.
3>Done building project "thirdparty_boost_context.vcxproj" -- FAILED.

2. project thirdparty_rocksbd:
1>D:\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.32.31326\include\nmmintrin.h(17,1): fatal  error C1189: #error:  This header is specific to X86 and X64 targets
1>Done building project "thirdparty_rocksdb.vcxproj" -- FAILED.

3. project osquery_tables_system_systemtable
199>D:\ARM_OsQuery\osquery\osquery\tables\system\cpuid.cpp(85,3): error C3861: '__cpuidex': identifier not found
199>Done building project "osquery_tables_system_systemtable.vcxproj" -- FAILED.

I need to use any other option for build? Thank you.

**Comment by pbo-linaro** [comment]
From latest [Actions CI ran on top of master](https://github.com/osquery/osquery/actions/runs/6316391805/job/17150900847), build is still successful.

I use those commands to build:
```
call "c:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Auxiliary/Build/vcvarsall.bat" x64_arm64
cmake -G Ninja -DCMAKE_BUILD_TYPE=Release -DOSQUERY_BUILD_TESTS=ON -DOSQUERY_BUILD_AWS=OFF ..
ninja
```

Are you using VS2019, and building from an arm64 VS prompt?

**Comment by lszcs90** [comment]
Thank you for reply.

VS2022 using, with 
`-- Selecting Windows SDK version 10.0.22621.0 to target Windows 10.0.19044.
-- MSVC toolset version: 143
-- osquery version: 5.9.1`

Tried to build with command line (directly) too and with opening sln in Visual Studio and to build there too.

For command line build I'm running the mentioned commands in an admin cmd.exe:
`cmake -G "Visual Studio 17 2022" -A arm64 .. -DOSQUERY_BUILD_AWS=OFF
cmake --build . --config RelWithDebInfo -j24`
For x64 build I'm using the same commands, but without the BUILD_AWS turned off.

**Comment by pbo-linaro** [comment]
I tried to build with VS2022, and I have some errors too.
So please build with VS2019.

**Comment by Smjert** [comment]
The issue is that you still need the toolchain file to properly set that the `CMAKE_SYSTEM_PROCESSOR` (which is used to determine the actual target) is `ARM64`.

So you need to create a file with:
```
set(CMAKE_SYSTEM_NAME Windows)
set(CMAKE_SYSTEM_PROCESSOR ARM64)
```

And then pass it as `-DCMAKE_TOOLCHAIN_FILE=<path>`; just be sure to remove the previously created `CMakeCache.txt` file in the build folder.

Also, be aware that the build might fail due to long paths not properly supported by msbuild; with this and the fact that msbuild is slower in building (due to bad parallelism) it's why we don't use it anymore.

Finally Visual Studio should be able to open a CMake project + Ninja directly.

**Comment by pbo-linaro** [comment]
I checked that VS2022 is correctly building osquery for windows-arm64, and that's the case too:
```
call "c:/Program Files/Microsoft Visual Studio/2022/Community/VC/Auxiliary/Build/vcvarsall.bat" x64_arm64
cmake -G Ninja -DCMAKE_BUILD_TYPE=Release -DOSQUERY_BUILD_TESTS=ON -DOSQUERY_BUILD_AWS=OFF ..
ninja
```

As @Smjert mentioned, avoid using `msbuild` when you can. In general, `ninja` based build is a much better solution overall.

**Comment by lszcs90** [comment]
Hello, 

Thank you for your answers. 
We managed to build it for ARM64 using the -DCMAKE_TOOLCHAIN_FILE option described by @Smjert . 
We used the following commands: 
```
cmake -G "Visual Studio 17 2022" -A arm64 .. -DOSQUERY_BUILD_AWS=OFF -DCMAKE_TOOLCHAIN_FILE=<pathToFile>
cmake --build . --config RelWithDebInfo -j24
```
and the toolchain file with the following content: 
```
set(CMAKE_SYSTEM_NAME Windows)
set(CMAKE_SYSTEM_PROCESSOR ARM64)
```

Maybe this could be added to the "Step 2: Download and build source on Windows" section from docs.

**Pull request merged**